src/Evolution/Systems/ScalarAdvection/TimeDerivativeTerms.cpp

```diff
@@ -16,8 +16,10 @@ namespace ScalarAdvection {
 template <size_t Dim>
 void TimeDerivativeTerms<Dim>::apply(
     gsl::not_null<Scalar<DataVector>*> /*non_flux_terms_dt_vars*/,
+    gsl::not_null<tnsr::I<DataVector, Dim>*> returned_velocity_field,
```

Comment:
can you mark all the `gsl::not_null` arguments as `const gsl::not_null`? (only in the function definition)

Reply:
Yep.
Actually this is a small question which I've wondered from a time ago: for all those `gsl::not_null` arguments, why do we need `const` in the function definition but not in function declarations and instantiations?

Reply:
ah, in the function definition the variable is used and so whether it's `const` or not matters (i.e. can you modify it?). In the declaration (and instantiations) they aren't used and the person calling the function doesn't need to know whether the variable (which is copied into the function, but it's just a pointer here) has its value changed inside the function. You don't want to modify the pointer, but its contents, so marking the pointer as `const` is good to catch bugs.

Is my explanation clear?

Reply:
Okay!

Then, are we _intentionally_ omitting `const` for some argument in declaration (and instantiations)? For instance, in the declaration below in the Rusanov.hpp
```
  static void dg_boundary_terms(
      gsl::not_null<Scalar<DataVector>*> boundary_correction_u,
      const Scalar<DataVector>& u_int,
      const Scalar<DataVector>& normal_dot_flux_u_int,
      const Scalar<DataVector>& abs_char_speed_int,
      const Scalar<DataVector>& u_ext,
      const Scalar<DataVector>& normal_dot_flux_u_ext,
      const Scalar<DataVector>& abs_char_speed_ext,
      dg::Formulation dg_formulation) noexcept;
```
I see no `const` for the first and last one while there are in the definition of this. Is there a special reason?

Reply:
the first and last arguments are taken by value, while the middle ones are by reference. When passing a reference the caller/declaration needs to know whether the argument is a reference, const reference, or value. For values it doesn't matter if they are const or not.

Reply:
Ah, I see! Thank you :smile: